apiserver/facades/client/machinemanager/machinemanager_test.go

```diff
@@ -330,6 +330,118 @@ func (s *MachineManagerSuite) TestUpgradeSeriesLocksMachine(c *gc.C) {
+			Series: "trusty",
+		},
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result.Error, gc.ErrorMatches, "machine 76 not found")
+}
+
+func (s *MachineManagerSuite) TestUpgradeSeriesPrepareNotMachineTag(c *gc.C) {
+	apiV5 := machinemanager.MachineManagerAPIV5{MachineManagerAPI: s.api}
+	unitTag := names.NewUnitTag("mysql/0")
+	result, err := apiV5.UpgradeSeriesPrepare(
+		params.UpdateSeriesArg{
+			Entity: params.Entity{
+				Tag: unitTag.String()},
+			Series: "trusty",
+		},
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result.Error, gc.ErrorMatches, "\"unit-mysql-0\" is not a valid machine tag")
+}
+
+func (s *MachineManagerSuite) TestUpgradeSeriesPreparePermissionDenied(c *gc.C) {
+	user := names.NewUserTag("fred")
+	s.setAPIUser(c, user)
+	apiV5 := machinemanager.MachineManagerAPIV5{MachineManagerAPI: s.api}
+	machineTag := names.NewMachineTag("0")
+	_, err := apiV5.UpgradeSeriesPrepare(
+		params.UpdateSeriesArg{
+			Entity: params.Entity{
+				Tag: machineTag.String()},
+			Series: "xenial",
+		},
+	)
+	c.Assert(err, gc.ErrorMatches, "permission denied")
+}
+
+func (s *MachineManagerSuite) TestUpgradeSeriesPrepareBlockedChanges(c *gc.C) {
+	apiV5 := machinemanager.MachineManagerAPIV5{MachineManagerAPI: s.api}
+	s.st.blockMsg = "TestUpgradeSeriesPrepareBlockedChanges"
+	s.st.block = state.ChangeBlock
+	_, err := apiV5.UpgradeSeriesPrepare(
+		params.UpdateSeriesArg{
+			Entity: params.Entity{
+				Tag: names.NewMachineTag("0").String()},
+			Series: "xenial",
+		},
+	)
+	c.Assert(params.IsCodeOperationBlocked(err), jc.IsTrue, gc.Commentf("error: %#v", err))
+	c.Assert(errors.Cause(err), jc.DeepEquals, &params.Error{
+		Message: "TestUpgradeSeriesPrepareBlockedChanges",
+		Code:    "operation is blocked",
+	})
+}
+
+func (s *MachineManagerSuite) TestUpgradeSeriesPrepareNoSeries(c *gc.C) {
+	apiV5 := machinemanager.MachineManagerAPIV5{MachineManagerAPI: s.api}
+	result, err := apiV5.UpgradeSeriesPrepare(
+		params.UpdateSeriesArg{
+			Entity: params.Entity{Tag: names.NewMachineTag("0").String()},
+		},
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result, jc.DeepEquals, params.ErrorResult{
+		Error: &params.Error{
+			Code:    params.CodeBadRequest,
+			Message: `series missing from args`,
+		},
+	})
+}
+
+func (s *MachineManagerSuite) TestUpgradeSeriesPrepareIncompatibleSeries(c *gc.C) {
+	s.setupUpdateMachineSeries(c)
+	s.st.machines["0"].SetErrors(&state.ErrIncompatibleSeries{[]string{"yakkety", "zesty"}, "xenial"})
+	apiV5 := machinemanager.MachineManagerAPIV5{MachineManagerAPI: s.api}
+	result, err := apiV5.UpgradeSeriesPrepare(
+		params.UpdateSeriesArg{
+			Entity: params.Entity{Tag: names.NewMachineTag("0").String()},
+			Series: "xenial",
+			Force:  false,
+		},
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result, jc.DeepEquals, params.ErrorResult{
+		Error: &params.Error{
+			Code:    params.CodeIncompatibleSeries,
+			Message: "series \"xenial\" not supported by charm, supported series are: yakkety,zesty",
+		},
+	})
```

Comment:
Of the tests that I see here, the only path that doesn't seem to be covered is if VerifyUnitsSeries fails. Is that being done separately?

Reply:
VerifyUnitsSeries() failure is covered by TestUpgradeSeriesPrepareIncompatibleSeries()